File: 04-FloatingPoint/FP.py
import numpy as np

def relative_error(measured, actual):
    if actual == 0:
        return 0.0
    return abs(measured - actual) / abs(actual) * 100

def fast_variance(data):
    n = len(data)
    if n < 1: return 0.0
    sum_x = np.sum(data)
    sum_x2 = np.sum(data**2)
    var = (sum_x2 - (sum_x**2) / n) / n
    return var

def two_pass_variance(data):
    n = len(data)
    if n < 1: return 0.0
    mean = np.mean(data)
    sum_diff2 = np.sum((data - mean)**2)
    return sum_diff2 / n

def one_pass_variance(data):
    n = 0
    mean = 0.0
    M2 = 0.0
    for x in data:
        n += 1
        delta = x - mean
        mean += delta / n
        delta2 = x - mean
        M2 += delta * delta2
    return M2 / n

def run_experiment(mu, sigma):
    np.random.seed(42)
    raw_data = np.random.normal(mu, sigma, 1000)
    
    data_high = raw_data.astype(np.float64)
        
    reference_var = one_pass_variance(data_high)
    
    results = {}
    for dtype in [np.float32, np.float64]:
        data = raw_data.astype(dtype)
        results[dtype.__name__] = {
            'Быстрый': fast_variance(data),
            'Двухпроходной': two_pass_variance(data),
            'Однопроходной': one_pass_variance(data)
        }
    
    return reference_var, results

samples = [(1, 1), (10, 0.1), (100, 0.01)]

print(f"{'Выборка (mu, sigma)':<20} | {'Тип':<10} | {'Метод':<15} | {'Отн. ошибка (%)':<15}")
print("-" * 70)

for mu, sigma in samples:
    ref, res_all = run_experiment(mu, sigma)
    
    for dtype_name, methods in res_all.items():
        first_in_dtype = True
        for method_name, val in methods.items():
            err = relative_error(val, ref)
            
            mu_sig_str = f"({mu:g}, {sigma:g})" if (first_in_dtype and dtype_name == 'float32') else ""
            dt_str = dtype_name if first_in_dtype else ""
            
            print(f"{mu_sig_str:<20} | {dt_str:<10} | {method_name:<15} | {err:15.10f}")
            first_in_dtype = False
        print("-" * 70)

File: 04-FloatingPoint/FP.cpp
#include <iomanip>
#include <iostream>
#include <random>

const int NUMBERS = 1000;
#define RESET "\033[0m"
#define CYAN "\033[36m"
#define MAGENTA "\033[35m"

template <typename T> double relative_error__(T measured, double actual) {
  if (actual == 0)
    return 0.0;
  return std::abs(static_cast<double>(measured) - actual) / std::abs(actual) *
         100.0;
}

template <typename T> T sum__(const std::vector<T> &data) {
  T sum = 0;
  for (const auto &x : data)
    sum += x;

  return sum;
}

template <typename T> T fast_variance(const std::vector<T> &data) {
  size_t n = data.size();
  if (n < 1)
    return 0;

  T sum_x = sum__(data);
  T sum_x_sqr = 0;
  for (const auto &x : data)
    sum_x_sqr += x * x;

  return (sum_x_sqr - (sum_x * sum_x) / static_cast<T>(n)) / static_cast<T>(n);
}

template <typename T> T two_pass_variance(const std::vector<T> &data) {
  size_t n = data.size();
  if (n < 1)
    return 0;

  T mean = sum__(data) / static_cast<T>(n);

  T sum_diff_sqr = 0;
  for (const auto &x : data)
    sum_diff_sqr += (x - mean) * (x - mean);

  return sum_diff_sqr / static_cast<T>(n);
}

template <typename T> T one_pass_variance(const std::vector<T> &data) {
  size_t n = 0;
  T mean = 0;
  T var = 0;

  for (const auto &x : data) {
    n++;
    if (n == 1) {
      mean = x;
      var = 0;
      continue;
    }

    T old_mean = mean;
    mean = old_mean + (x - old_mean) / static_cast<T>(n);

    T term1 = (x - old_mean) * (x - mean);
    var = var + (term1 - var) / static_cast<T>(n);
  }

  return var;
}

void run_test(const std::pair<int, float> &&sample) {
  auto [mu, sigma] = sample;
  std::cout << CYAN << "Тест: mu = " << mu
            << ", sigma = " << std::setprecision(2) << sigma
            << ", n = " << NUMBERS << ";" << MAGENTA
            << " D[X]_теор = " << std::setprecision(4) << sigma * sigma << RESET
            << "\n";

  double theorVar = sigma * sigma;

  std::mt19937 gen(42);
  std::normal_distribution<double> dist(mu, sigma);

  std::vector<float> data32;
  std::vector<double> data64;
  for (int i = 0; i < NUMBERS; ++i) {
    double val = dist(gen);
    data32.push_back(static_cast<float>(val));
    data64.push_back(val);
  }

  auto print_row = [&](std::string type, std::string method, double res,
                       double ref) {
    std::cout << std::left << std::setw(10) << type << " | " << std::setw(15)
              << method << " | " << std::fixed << std::setprecision(9)
              << std::setw(9) << res << (res < 0 ? " (!)" : "     ") << " | "
              << std::setprecision(6) << relative_error__(res, ref)
              << std::endl;
  };

  std::cout << std::left << std::setw(10) << "    Тип" << "    | "
            << std::setw(15) << "    Метод" << "    | " << std::setw(14)
            << "       D[X]" << "   | "
            << "Отн. ошибка (%)" << std::endl;
  std::cout << std::string(62, '-') << std::endl;

  // float32
  print_row("float32", "Быстрый      ", fast_variance(data32), theorVar);
  print_row("float32", "Двухпроходной", two_pass_variance(data32), theorVar);
  print_row("float32", "Однопроходной", one_pass_variance(data32), theorVar);

  std::cout << std::string(62, '-') << std::endl;

  // float64
  print_row("float64", "Быстрый      ", fast_variance(data64), theorVar);
  print_row("float64", "Двухпроходной", two_pass_variance(data64), theorVar);
  print_row("float64", "Однопроходной", one_pass_variance(data64), theorVar);

  std::cout << "\n\n";
}

int main() {
  std::vector<std::pair<int, float>> samples{{1, 1}, {10, 0.1}, {100, 0.01}};

  for (const auto &sample : samples)
    run_test(std::move(sample));

  // Тест, в котором наблюдается отрицательное значение дисперсии. Здесь среднее
  // большое, а дисперсия маленькая
  // run_test({10000, 0.01f});
}
